pep-0613.rst

```diff
@@ -101,10 +102,11 @@ Scope Restrictions:
     x = ClassName
 
 The outer :code:`x` is a valid type alias, but type checkers must error if the
-inner :code:`x` is ever used as a type. This is confusing because the alias
-declaration rule is not explicit, and the error is not thrown on the location
-of the inner type alias declaration but rather on every one of its subsequent
-use cases.
+inner :code:`x` is ever used as a type because type aliases cannot be defined
+inside a nested scope.
+This is confusing because the alias declaration rule is not explicit, and because
+a type error will not be thrown on the location of the inner type alias declaration
+but rather on every one of its subsequent use cases.
```

Comment:
I just realize something (about the example below, where I can't attach a comment): why don't we allow `x: TypeAlias = ClassName` inside a function? It's entirely unambiguous. (The reason may be that it's complicated for type checkers to handle type aliases with limited scope, although it would seem that nested class declarations work just fine.)

Reply:
At the moment MyPy and Pyre both disallow nested aliases, I assume b/c getting them right is equivalent to getting nested classes as types right within the scope of the parent class, but is a little more complicated otherwise.

Due to type variable scoping, it's not always sound to be able to reference a nested type alias outside of its scope using qualification, ie:
```
class Foo(Generic[_T]):
    MyType: TypeAlias = List[_T]

def foo() -> Foo.MyType: ...
```

Pyre currently doesn't have a system for scoping types (ie., disallowing the bad usage above), and I think MyPy and Pyre actually have identical behavior at the moment, where nested aliases are not supported at all, and nested classes can be referred to in-scope as well as with qualification outside of the parent class. Are nested aliases something we want to support?

Reply:
Mypy doesn't understand the `: TypeAlias` notation yet; without that, it gives an accurate error on your example, I think:
```
class Foo(Generic[_T]):
    MyType = List[_T]  # E: Can't use bound type variable "_T" to define generic alias
```
(So it seems to understand that the intention is a type alias, and it realizes there is a problem with it.)

But I'm happy not to open this can of worms, so I agree with your current wording. I'll land the PR and then I'll let the SC know.